state/factory/workingset.go

```diff
@@ -227,3 +223,248 @@ func (ws *workingSet) DelState(opts ...protocol.StateOption) (uint64, error) {
 	}
 	return ws.height, ws.delStateFunc(cfg.Namespace, cfg.Key)
 }
+
+// createGenesisStates initialize the genesis states
+func (ws *workingSet) CreateGenesisStates(ctx context.Context) error {
+	if bcCtx, ok := protocol.GetBlockchainCtx(ctx); ok {
+		for _, p := range bcCtx.Registry.All() {
+			if gsc, ok := p.(protocol.GenesisStateCreator); ok {
+				if err := gsc.CreateGenesisStates(ctx, ws); err != nil {
+					return errors.Wrap(err, "failed to create genesis states for protocol")
+				}
+			}
+		}
+	}
+
+	return ws.finalize()
+}
+
+func (ws *workingSet) validateNonce(blk *block.Block) error {
+	accountNonceMap := make(map[string][]uint64)
+	for _, selp := range blk.Actions {
+		caller, err := address.FromBytes(selp.SrcPubkey().Hash())
+		if err != nil {
+			return err
+		}
+		appendActionIndex(accountNonceMap, caller.String(), selp.Nonce())
+	}
+
+	// Special handling for genesis block
+	if blk.Height() == 0 {
+		return nil
+	}
+	// Verify each account's Nonce
+	for srcAddr, receivedNonces := range accountNonceMap {
+		confirmedState, err := accountutil.AccountState(ws, srcAddr)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get the confirmed nonce of address %s", srcAddr)
+		}
+		receivedNonces := receivedNonces
+		sort.Slice(receivedNonces, func(i, j int) bool { return receivedNonces[i] < receivedNonces[j] })
+		for i, nonce := range receivedNonces {
+			if nonce != confirmedState.Nonce+uint64(i+1) {
+				return errors.Wrapf(
+					action.ErrNonce,
+					"the %d nonce %d of address %s (confirmed nonce %d) is not continuously increasing",
+					i,
+					nonce,
+					srcAddr,
+					confirmedState.Nonce,
+				)
+			}
+		}
+	}
+	return nil
+}
+
+func (ws *workingSet) Process(ctx context.Context, actions []action.SealedEnvelope) ([]*action.Receipt, error) {
```

Comment:
why we need both Process() and process(), they are identical func?
i mean, there is no lock/unlock in Process()